opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -2063,38 +2064,104 @@ namespace Opm {
+                break;
+            }
+            case GConSale::MaxProcedure::WELL: {
+                OPM_DEFLOG_THROW(std::runtime_error, "Group " + group.name() + "GCONSALE exceed limit WELL not implemented", deferred_logger);
+                break;
+            }
+            case GConSale::MaxProcedure::PLUG: {
+                OPM_DEFLOG_THROW(std::runtime_error, "Group " + group.name() + "GCONSALE exceed limit PLUG not implemented", deferred_logger);
+                break;
+            }
+            case GConSale::MaxProcedure::MAXR: {
+                OPM_DEFLOG_THROW(std::runtime_error, "Group " + group.name() + "GCONSALE exceed limit MAXR not implemented", deferred_logger);
+                break;
+            }
+            case GConSale::MaxProcedure::END: {
+                OPM_DEFLOG_THROW(std::runtime_error, "Group " + group.name() + "GCONSALE exceed limit END not implemented", deferred_logger);
+                break;
+            }
+            case GConSale::MaxProcedure::RATE: {
+                    well_state.setCurrentProductionGroupControl(group.name(), Group::ProductionCMode::GRAT);
+                    ss << "Maximum GCONSALE limit violated for " << group.name() << ". The group is switched from " << Group::ProductionCMode2String(oldProductionControl) << " to ";
+                    ss << Group::ProductionCMode2String(Group::ProductionCMode::GRAT) << " and limited by the maximum sales rate after consumption and import are considered" ;
+                    well_state.setCurrentGroupGratTargetFromSales(group.name(), production_target);
+                break;
+            }
+            default:
+                throw("Invalid procedure for maximum rate limit selected for group" + group.name());
+            }
         }
         if (sales_rate < gconsale.min_sales_rate) {
-            OPM_THROW(std::runtime_error, "Group " + group.name() + " has sale rate less then minimum permitted value. Not implemented in Flow" );
+            const Group::ProductionCMode& currentProductionControl = well_state.currentProductionGroupControl(group.name());
+            if ( currentProductionControl == Group::ProductionCMode::GRAT ) {
+                ss << "Group " + group.name() + " has sale rate less then minimum permitted value and is under GRAT control. \n";
+                ss << "The GRAT is increased to meet the sales minimum rate. \n";
+                well_state.setCurrentGroupGratTargetFromSales(group.name(), production_target);
+            //} else if () {//TODO add action for WGASPROD
+            //} else if () {//TODO add action for drilling queue
+            } else {
+                ss << "Group " + group.name() + " has sale rate less then minimum permitted value but cannot increase the group production rate, \n";
+                ss << "adjust gas production using WGASPROD nor drill new wells to meat the sales target. No action is taken. \n ";
```

Comment:
Do we support WGASPROD? Also, 'nor'->'or' and 'meat'->'meet'.

Reply:
We don't support WGASPROD.

Reply:
Then I think we should not recommend it as a solution?

Reply:
It is , not a . So the meaning of the text is that we can not adjust gas production using WGASPROD, etc. I can add something like. "Not supported in Flow. " to clearly this.

Reply:
Ah, I misunderstood a little then. Good to have a clearer message.